selfhost/typechecker.jakt

```diff
@@ -1460,6 +1548,334 @@ struct Typechecker {
         }
     }
 
+    function statement_definitely_returns(this, anon statement: CheckedStatement) -> bool => match statement{
+        Return(expr) => true
+        // FIXME: CheckedStatement::If branches ugly implementation due to current compiler limitations
+        If(guard, then_block, else_statement) => {
+            mut ret = false
+            let guard_result = match guard {
+                Boolean(val, span) => {
+                    if val and then_block.definitely_returns {
+                        ret = true
+                    }
+                    yield ret
+                }
+                else => false
+            }
+            if not guard_result and else_statement.has_value(){
+                if then_block.definitely_returns and .statement_definitely_returns(else_statement.value()) {
+                    ret = true
+                }
+            }
+            yield ret
+        }
+        Block(block) => block.definitely_returns
+        Loop(block) => block.definitely_returns
+        While(guard, block) => block.definitely_returns
+        Expression(match_expr) => match match_expr {
+            Match(expr, match_cases, span, type_id, all_variants_constant) => {
+                yield checked_expression_definitely_returns(match_expr)
+            }
+            else => false
+        }
+        else => false
+    }
+
+    // FIXME: Use [TypeId: TypeID] without TypeId.to_string()/TypeId::from_string() workaround
+    function check_types_for_compat(mut this, lhs_type_id: TypeId, rhs_type_id: TypeId, mut generic_inferences: [String:String], span: Span) throws -> bool {
+        mut ret = true
+        let lhs_type = .get_type(lhs_type_id)
+
+        let lhs_type_id_string = lhs_type_id.to_string()
+        let rhs_type_id_string = rhs_type_id.to_string()
+
+        let optional_struct_id = .find_struct_in_prelude("Optional")
+        let weakptr_struct_id = .find_struct_in_prelude("WeakPtr")
+
+        // TODO: Skip the type compatibility check if assigning a T to a T? or to a
+        //       weak T? without going through `Some`.
+
+        match lhs_type {
+            TypeVariable(name) => {
+                let seen_type_id_string = generic_inferences.get(lhs_type_id_string)
+                if seen_type_id_string.has_value() {
+                    if seen_type_id_string.value() != rhs_type_id_string {
+                        .error(
+                            format(
+                                "Type mismatch: expected `{}` but got `{}`"
+                                .type_name(TypeId::from_string(seen_type_id_string.value()))
+                                .type_name(rhs_type_id)
+                            )
+                            span
+                        )
```

Comment:
I won't write this under every `.error(...)`, but I think you have to explicitly `return false` or set `ret = false` after each error (personally I would remove `ret` and just `return true` at the end if the checks did not fail along the way).
The original Rust code used return type `Option<JaktError>`, so it was enough to emit an error anywhere a check failed, but since the Jakt function here returns a boolean, you have to return false in addition to emitting the error.